compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/PrefetchSharedMemoryCopy.cpp

```diff
@@ -176,9 +176,54 @@ class LoopPrefetcher {
     });
   }
 
+  void getValueDependenciesForIf(scf::IfOp ifOp,
+                                 DenseSet<Operation *> &readDependencies,
+                                 DenseSet<Operation *> &writeDependencies) {
+    // scf.if with results should be supported directly through the usual
+    // handling so bail-out here in that case
+    if (ifOp->getNumResults() != 0)
+      return;
+    bool hasGlobalRead = false;
+    bool hasSharedWrite = false;
+    // Else region not yet supported.
+    if (!ifOp.getElseRegion().empty()) {
+      return;
+    }
+    ifOp->walk([&](Operation *op) {
+      if (auto readOp = dyn_cast<vector::TransferReadOp>(op)) {
+        auto sourceType = dyn_cast<MemRefType>(readOp.getBase().getType());
+        if (hasGlobalMemoryAddressSpace(sourceType)) {
+          hasGlobalRead = true;
+        }
+      }
+      if (auto writeOp = dyn_cast<vector::TransferWriteOp>(op)) {
+        auto dstType = dyn_cast<MemRefType>(writeOp.getBase().getType());
+        if (hasSharedMemoryAddressSpace(dstType)) {
+          hasSharedWrite = true;
+        }
+      }
+    });
+    // if op has both read and write stages and hence we cannot do prefetching.
+    if (hasGlobalRead && hasSharedWrite) {
+      return;
+    }
+    if (hasSharedWrite) {
+      getValueDependencies(ifOp, writeDependencies, /*noTransferReads=*/true);
+      return;
+    }
+    // By default we can mark the if op as read stage.
+    // TODO (nirvedhmeshram) : Add handling for private memory allocations. e.g
+    // write to private memory could go in write stage. But the analysis
+    // in `getValueDependencies` also needs to be aware of private memory for
+    // this so that needs to be added at the same time.
```

Comment:
Just for information, is this the collaboration work we are about to land? That we categorize write to private to belong to write stage too.

Reply:
yes